File: Easy/valid_parenthesis.py
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

# An input string is valid if:
# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
# Every close bracket has a corresponding open bracket of the same type.
 
# Example 1:
# Input: s = "()"
# Output: true

# Example 2:
# Input: s = "([]){}"
# Output: true

# Example 3:
# Input: s = "(]"
# Output: false

# ---------------------- My Solution ----------------------------------


class Solution:
    def isValid(self, s: str) -> bool:
        bracketCloser = {
            "(": ")",
            "{": "}",
            "[": "]"
        }

        stack = []

        for character in s:
            if character == '(' or character == "{" or character == "[":
                stack.append(bracketCloser[character])
            else:
                if len(stack) == 0: 
                    return False
                if character == stack[len(stack)-1]:
                    stack.pop()
                else:
                    return False; 
        
        if len(stack) == 0: 
            return True
        else: 
            return False


        

File: Easy/valid_parenthesis.ts
// You are given a string s consisting of the following characters: '(', ')', '{', '}', '[' and ']'.

// The input string s is valid if and only if:

// Every open bracket is closed by the same type of close bracket.
// Open brackets are closed in the correct order.
// Every close bracket has a corresponding open bracket of the same type.
// Return true if s is a valid string, and false otherwise.

// Example 1:

// Input: s = "[]"

// Output: true
// Example 2:

// Input: s = "([{}])"

// Output: true
// Example 3:

// Input: s = "[(])"

// Output: false
// Explanation: The brackets are not closed in the correct order.

// Constraints:

// 1 <= s.length <= 1000

// ____________________________________________ My Solution ______________________________________________


function isValid(s: string): boolean {
        
          const myStack : string[] = []
          const matchMap : { [key : string]: string } = 
          { ")" : "(", 
            "]" : "[",
            "}" : "{" }
       
             for (let i = 0; i< s.length; i++) {
                   
                   if (s[i] == "(" || s[i] == "{" || s[i] == "[") {
                       myStack.unshift(s[i])
                   }
                   else {
                        
                        if (myStack[0] == matchMap[s[i]] ) {
                            myStack.shift(); 
                        }
                        else {
                            return false
                        }
                   }
             }

             if (myStack.length != 0) {
                return false 
             }

           return true 
    }
